soc: st: stm32wba: add support for sys_poweroff

ivanwagner opened the pull request:
Adding sys_poweroff to WBA family. It switches to standby (non-retention) mode. It can be woken up by any wakeup source.

It is a contribution to #81119 for the WBA family.

ivanwagner commented on soc/st/stm32/stm32wbax/poweroff.c:
```diff
@@ -0,0 +1,30 @@
+	LL_PWR_ClearFlag_WU();
+#endif /* CONFIG_STM32_WKUP_PINS */
+
+	LL_PWR_SetPowerMode(LL_PWR_MODE_STANDBY);
```
The U5 standby used with PM_STATE_STANDBY is with ram-retention or not?

ivanwagner replied:
[Here](https://github.com/zephyrproject-rtos/zephyr/issues/81119#issuecomment-2464408410 ) you mention that currently wakeup pins only work if CONFIG_POWEROFF is activated. It means that for the time being the WBA family could not wakeup for the lack of poweroff.c (which this PR adds). In my opinion it is a limitation. In my use case and with this simple implementation I can reach the min consumption and wakeup and reboot without any other PM features. So, if in the future this same behavior would be added via PM_STATE_STANDBY I just see it as a plus (I'm looking at the issue from a high level point of view). By the way, at the moment it seems that PM is broken within the WBA family...

mathieuchopstm replied:
> > I don't see any issue with Standby (no retention) being made available under CONFIG_POWEROFF.
> 
> But why not as PM_STATE_STANDBY?

Please read https://docs.zephyrproject.org/latest/doxygen/html/group__subsys__pm__states.html#ga20e2f5ea9027a3653e5b9cc5aa1e21d5

mathieuchopstm replied:
> What part exactly?

> `PM_STATE_STANDBY`
> [...] it should be the same state with _[sic]_ suspend to idle state on uniprocessor system.

WBA is a uniprocessor series, so `PM_STATE_STANDBY` should not be used but instead:
> `PM_STATE_SUSPEND_TO_IDLE`
> The system goes through a normal platform suspend where it puts all of the cores in deepest possible idle state and *may* puts _[sic]_ peripherals into low-power states. **No operating state is lost** (ie. **the cpu core does not lose execution context**)

Emphases mine. Under no reasonable interpretation would STM32 Standby fall in this category (be it `Standby` (no retention) or `Standby retention`).

STM32WBA `Standby retention` should be implemented as `PM_STATE_SUSPEND_TO_RAM` (there is already an ongoing effort for this)

If anything, `Standby` (no retention) would be `PM_STATE_SOFT_OFF`, which is quite literally a "powered off" state...
(There is still some usage of `PM_STATE_SOFT_OFF` in tree, but I would be willing to bet it is merely legacy from a pre-`CONFIG_POWEROFF` time)

Oleh-Kravchenko replied:
> [@mathieuchopstm](https://github.com/mathieuchopstm) thanks for the clarification, I see the problem now.
> 
> > (There is still some usage of PM_STATE_SOFT_OFF in tree, but I would be willing to bet it is merely legacy from a pre-CONFIG_POWEROFF time)
> 
> Poweroff is orthogonal to PM power states, it's invoked directly by the user, whereas power states are selected by the PM subsystem....

Yes, that looks not very solid.
sys_power_off() vs PM_STATE_SOFT_OFF

mathieuchopstm replied:
> > (There is still some usage of PM_STATE_SOFT_OFF in tree, but I would be willing to bet it is merely legacy from a pre-CONFIG_POWEROFF time)
> 
> Poweroff is orthogonal to PM power states, it's invoked directly by the user, whereas power states are selected by the PM subsystem....

Something something https://github.com/zephyrproject-rtos/zephyr/commit/90b247b90b1a8f8b545a325a303ff5f52bceae80

mathieuchopstm commented on dts/arm/st/wba/stm32wba.dtsi:
```diff
@@ -156,6 +156,67 @@
+			#size-cells = <0>;
+
+			wkup-pin@1 {
+				reg = <0x1>;
+				wkup-gpios = <&gpioa 0 STM32_PWR_WKUP_EVT_SRC_0>,
+							 <&gpiob 2 STM32_PWR_WKUP_EVT_SRC_1>;
+			};
```
```suggestion
				reg = <0x1>;
				wkup-gpios = <&gpioa 0 STM32_PWR_WKUP_EVT_SRC_0>,
					     <&gpiob 2 STM32_PWR_WKUP_EVT_SRC_1>;
			};
```
Ditto for all instances.

ivanwagner commented on dts/arm/st/wba/stm32wba.dtsi:
```diff
@@ -156,6 +156,67 @@
+			wkup-pin@1 {
+				reg = <0x1>;
+				wkup-gpios = <&gpioa 0 STM32_PWR_WKUP_EVT_SRC_0>,
+							<&gpiob 2 STM32_PWR_WKUP_EVT_SRC_1>;
```
I think it's ok now...

mathieuchopstm reviewed (approved):
LGTM.

mathieuchopstm reviewed (changes requested):
Something I almost missed (sorry) - please split changes over three commits:
* `dts: arm: st: stm32wba: ...` for DTSI changes
* `soc: st: stm32wba: ...` for SoC code changes/addition
* `samples: boards: st: wkup_pins: add overlay for Nucleo-WBA55CG` for overlay

ivanwagner committed:
dts: arm: st: stm32wba: added power control peripheral

Added support for wakeup pins (events).

Signed-off-by: Ivan Wagner <ivan.wagner@tecinvent.ch>

ivanwagner committed:
soc: st: stm32wba: add support for sys_poweroff

Adding sys_poweroff to WBA family. It switches to standby
(non-retention) mode. It can be woken up by any wakeup source.

Signed-off-by: Ivan Wagner <ivan.wagner@tecinvent.ch>

ivanwagner committed:
samples: boards: st: wkup_pins: add overlay for Nucleo-WBA55CG

Added wakeup pins sample for nucleo-wba55cg board.

Signed-off-by: Ivan Wagner <ivan.wagner@tecinvent.ch>

JarmouniA reviewed (commented):
Will +1 once you add something to this effect in the wake-up pins sample doc (I don't see another place more appropriate for it in the doc)

_Note that Zephyr Poweroff is mapped to STM32 low-power mode with the lowest power consumption, which is either Standby or Shutdown depending on the SoC series._

ivanwagner commented:
@JarmouniA ok I added the comment in the samples README file.

ivanwagner commented:
@erwango What do you think about it?

mathieuchopstm commented:
> @erwango What do you think about it?

Erwan is out-of-office so he can't reply right now...

etienne-lms reviewed (commented):
Some nitpicking comments, otherwise LGTM.

ivanwagner commented on dts/arm/st/wba/stm32wba.dtsi:
```diff
@@ -156,6 +156,67 @@
 
+		pwr: power@46020800 {
+			compatible = "st,stm32-pwr";
+			reg = <0x46020800 0x400>; /* PWR register bank */
```
Yes I got it from there ;)

ivanwagner commented on dts/arm/st/wba/stm32wba.dtsi:
```diff
@@ -156,6 +156,66 @@
+			status = "disabled";
+
+			wkup-pins-nb = <8>; /* 8 system wake-up pins */
+			wkup-pin-srcs = <3>; /* 3 gpio sources associated with each wkup pin */
```
This is the table from the STM32WB5x ref manual (RM0493). There are three wakeup pin sources.

<img width="814" height="594" alt="image" src="https://github.com/user-attachments/assets/91d7e9b2-183b-4e5f-ab9c-8b6d8d6b8bf3" />

Am I missing something?

etienne-lms commented on soc/st/stm32/stm32wbax/poweroff.c:
```diff
@@ -0,0 +1,31 @@
+
+void z_sys_poweroff(void)
+{
+	stm32_pwr_wkup_pin_cfg_pupd();
```
@JarmouniA, @ivanwagner, FYI I've created https://github.com/zephyrproject-rtos/zephyr/pull/94444 to address this point, extended to other STM32 SoCs where applicable.

etienne-lms commented on soc/st/stm32/stm32wbax/poweroff.c:
```diff
@@ -0,0 +1,33 @@
+	LL_PWR_SetSRAM2SBRetention(LL_PWR_SRAM2_SB_NO_RETENTION);
+	LL_LPM_EnableDeepSleep();
+
+	__WFI();
```
I missed that: prefer `k_cpu_idle();` for consistency with other STM32 SoCs implementation?

etienne-lms replied:
ok

etienne-lms commented on dts/arm/st/wba/stm32wba.dtsi:
```diff
@@ -156,6 +156,66 @@
+		pwr: power@46020800 {
+			compatible = "st,stm32-pwr";
+			reg = <0x46020800 0x400>;
+			status = "disabled";
```
A recommendation for DTS coding style is to have the `status` property last in the node properties.
For consistency with the other nodes implementation, could you move it down?

etienne-lms replied:
It's not very consistent regarding the other nodes in this DTSI file (aside the adc node) but I must admit it's consistent with the `pwr` nodes implementation in the other STM32 SoCs DTSI files.

etienne-lms commented on soc/st/stm32/stm32wbax/poweroff.c:
```diff
@@ -0,0 +1,33 @@
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/kernel.h>
```
I think *kernel.h* is no more needed.

etienne-lms reviewed (commented):
Some nitpicking comments, otherwise LGTM.
